specs/Gem.spec

```diff
@@ -0,0 +1,211 @@
+rule burnUnderflowSpec(address recip, uint amt) {
+    
+    env e;
+    require(balanceOf(recip) < amt);
+
+    burn@withrevert(e, recip, amt);
+    assert lastReverted, "burn must revert if total supply underflows";
+}
+
+rule mintAndBurnWardSpec(address other, uint amt) {
+
+    env e;
+    address sender = e.msg.sender;
+    ward(e, sender, false);
+
+    mint@withrevert(e, other, amt);
+    assert lastReverted, "mint did not revert with non-ward sender";
+
+    burn@withrevert(e, other, amt);
+    assert lastReverted, "burn did not revert with non-ward sender";
+}
+
+
+rule totalSupplySpec(address target, uint amt) {
+
+    env e; method f; calldataarg args;
+    mathint total_supply_before = totalSupply();
+
+    f@withrevert(e, args);
+
+    mathint total_supply_after = totalSupply();
+
+    assert total_supply_before != total_supply_after 
+        => 
+    (f.selector == mint(address,uint).selector || f.selector == burn(address,uint).selector)
+    && wards(e.msg.sender) == true,
+    "function other than warded mint or burn changed total supply!";
+}
+
+rule wardSpec(address other_ward) {
+        
+    env e;
+    address sender = e.msg.sender;
+    require(wards(sender) == true); // for this spec, assume sender is already a ward
+
+    ward(e, other_ward, true);
+    assert wards(other_ward) == true;
+
+    ward(e, other_ward, false);
+    assert wards(other_ward) == false;
+
+}
+
+rule approveSpec(address spender, uint amt) {
+    
+    env e;
+    approve(e, spender, amt);
+    assert allowance(e.msg.sender, spender) == amt, "spender allowance does not match intended amount";
+}
+
+rule transferFromSpec(address src, address dst, uint amt) {
+
+    env e;
+    address sender = e.msg.sender;
+    require(allowance(src, sender) >= amt);
+    require(balanceOf(src) >= amt);
+    require(balanceOf(src) <= max_uint256);
+    require(balanceOf(dst) <= max_uint256 - amt);
+    
+    mathint balance_src_before = balanceOf(src);
+    mathint balance_dst_before = balanceOf(dst);
+    mathint allowance_before = allowance(src, sender);
+
+    transferFrom(e, src, dst, amt);
+
+    // true to matter what
+    assert allowance(src, sender) == allowance_before - amt 
+            || (allowance_before == max_uint256 && allowance_before == allowance(src, sender)), 
+        "allowance did not decrease by transferFrom amt";
+
+    // conditionals
+    if(src != dst){
+        assert balanceOf(src) == balance_src_before - amt, "src balance did not decrease by transferFrom amt";
+        assert balanceOf(dst) == balance_dst_before + amt, "dst balance did not increase by transferFrom amt";
+    } else {
+        assert balanceOf(src) == balanceOf(dst) 
+            && balanceOf(src) == balance_src_before
+            && balanceOf(dst) == balance_dst_before
+            && balance_dst_before == balance_src_before, 
+            "balances changed despite transferring to self";
+    }
+}
+
+rule transferFromRevertSpec(address src, address dst, uint amt) {
```

Comment:
`transferFromRevertsOnInsufficientBalanceOrAllowance`, we can go ham with some of these names